toolchain/parse/handle_decl_scope_loop.cpp

```diff
@@ -7,23 +7,33 @@
 namespace Carbon::Parse {
 
 // Handles an unrecognized declaration, adding an error node.
-static auto HandleUnrecognizedDecl(Context& context) -> void {
+static auto HandleUnrecognizedDecl(Context& context, int32_t subtree_start)
+    -> void {
   CARBON_DIAGNOSTIC(UnrecognizedDecl, Error,
                     "Unrecognized declaration introducer.");
   context.emitter().Emit(*context.position(), UnrecognizedDecl);
   auto cursor = *context.position();
-  auto semi = context.SkipPastLikelyEnd(cursor);
-  // Locate the EmptyDecl at the semi when found, but use the
-  // original cursor location for an error when not.
-  context.AddLeafNode(NodeKind::EmptyDecl, semi ? *semi : cursor,
-                      /*has_error=*/true);
+  context.SkipPastLikelyEnd(cursor);
+  auto iter = context.position();
+  --iter;
+  // Output a single invalid parse node if consumed a single token in this
+  // error, otherwise output an invalid parse subtree.
+  if (*iter == context.tree().node_token(Node(subtree_start))) {
+    context.ReplacePlaceholderNode(subtree_start, NodeKind::InvalidParse,
+                                   cursor, /*has_error=*/true);
+  } else {
+    context.ReplacePlaceholderNode(subtree_start, NodeKind::InvalidParseStart,
+                                   cursor, /*has_error=*/true);
+    // Mark everything up to the last token consumed as invalid.
+    context.AddNode(NodeKind::InvalidParseSubtree, *iter, subtree_start,
+                    /*has_error=*/true);
+  }
```

Comment:
Had you considered uniformly outputting a tree here? It'd reduce conditional behaviors in the output parse tree, also removing the math on `context.position()` which I wouldn't expect in general.

Also, why are you discarding the result of SkipPastLikelyEnd? I think that the prior selection of semi or cursor is consistent with the roots on declarations (valid and invalid), leading to this structure.

Reply:
I've added a comment about ignoring the return value of `SkipPastLikelyEnd`.

I could always output a subtree error, but it seems like that would be surprising to consumers if it represents a single token.